src/rust/ide/src/controller/visualization.rs

```diff
@@ -0,0 +1,254 @@
+#[derive(Debug,Fail)]
+#[allow(missing_docs)]
+pub enum VisualizationError {
+    #[fail(display = "Visualization \"{}\" not found.", identifier)]
+    NotFound {
+        identifier : VisualizationIdentifier
+    },
+    #[fail(display = "JavaScript visualization \"{}\" failed to compile.", identifier)]
+    CompileError {
+        identifier : VisualizationIdentifier
+    }
+}
+
+
+
+// ===============================
+// === VisualizationIdentifier ===
+// ===============================
+
+/// This enum is used to identify a visualization both in the project folder or natively embedded in
+/// IDE.
+#[derive(Clone,Debug,Display,Eq,PartialEq)]
+#[allow(missing_docs)]
+pub enum VisualizationIdentifier {
+    Embedded(String),
+    File(language_server::Path)
+}
+
+
+
+// ==============================
+// === EmbeddedVisualizations ===
+// ==============================
+
+#[allow(missing_docs)]
+pub type EmbeddedVisualizationName   = String;
+
+/// Embedded visualizations mapped from name to source code.
+#[derive(Shrinkwrap,Debug,Clone,Default)]
+#[shrinkwrap(mutable)]
+pub struct EmbeddedVisualizations {
+    #[allow(missing_docs)]
+    pub map : HashMap<EmbeddedVisualizationName,Rc<class::Handle>>
+}
+
+
+
+
+// ==============
+// === Handle ===
+// ==============
+
+const VISUALISATION_FOLDER : &str = "visualization";
+
+#[derive(Debug)]
+struct VisualizationControllerData {
+    language_server_rpc     : Rc<language_server::Connection>,
+    embedded_visualizations : EmbeddedVisualizations
+}
+
+/// Visualization Controller's state.
+#[derive(Debug,Clone,CloneRef)]
+pub struct Handle {
+    data : Rc<RefCell<VisualizationControllerData>>
+}
+
+impl Handle {
+    /// Creates a new visualization controller.
+    pub fn new
+    ( language_server_rpc     : Rc<language_server::Connection>
+    , embedded_visualizations : EmbeddedVisualizations) -> Self {
+        let data = VisualizationControllerData{language_server_rpc,embedded_visualizations};
+        let data = Rc::new(RefCell::new(data));
+        Self {data}
+    }
+
+    async fn list_file_visualizations(&self) -> FallibleResult<Vec<VisualizationIdentifier>> {
+        let data                = self.data.borrow();
+        let language_server_rpc = &data.language_server_rpc;
+        let root_id             = language_server_rpc.content_root();
+        let path                = language_server::Path::new(root_id,&[VISUALISATION_FOLDER]);
+        let file_list           = language_server_rpc.file_list(&path).await?;
+        let result              = file_list.paths.iter().filter_map(|object| {
+            if let language_server::FileSystemObject::File{name,path} = object {
+                let mut path = path.clone();
+                path.segments.push(name.to_string());
+                let root_id    = path.root_id;
+                let segments   = path.segments;
+                let path       = language_server::Path{root_id,segments};
```

Comment:
We should make `From<FileSystemObject> for Path`.
This logic is not specific to visulization controller.